### DISCUSSION TITLE
Refresh route respond : Invalid user credentials

### DISCUSSION THREAD

**Discussion by MrCasquette** [directus/directus#14109] [Archive]
### Describe the Bug

I'm not quite sure of myself cause i'm junior, but it's seems Impossible to refresh the token. Thank you for helping me :)

### To Reproduce

I've successfully logged in and saved token and refresh in localstorage. But when i send the refresh saved there in credentials on auth/refresh REST route the error is still the same... :(

### Errors Shown

401 Unauthorized : Invalid user credentials.

### What version of Directus are you using?

9.12.2

### What version of Node.js are you using?

v12.22.5

### What database are you using?

Mysql

### What browser are you using?

Chrome

### How are you deploying Directus?

Self-Hosted on Ubuntu

**Comment by azrikahar** [comment]
Can you share your exact setup for this? Perhaps what is the framework (if any) you're using? 

Certain frameworks that have a mix of client side and server side renderings has been one of the main source of confusion when users are using the Directus SDK 🙂

**Reply by MrCasquette** [reply]
Thank you for the reply and sorry for opening an issue instead of a discussion. 
I don't use SDK. I try to use directus on my own with personnal async function. Hosted on my personnal VPS.
Beacause the SDK seems a bit complicated to me. I use React in front with a create-react-app template and axios for the request.
At the moment everything seems to work but the refresh route.
Because I'm newbie feel free to start from the beginning or say something you think evident or stupid, that won't be to me. And if you need some more information please ask, maybe I forget. Thank you again. 😃 I Love Directus ❤️

**Reply by irishburlybear** [reply]
Ok, So I am getting this error. My setup: I'm using sveltekit. Now, if I refresh from the server side(endpoint) I can get refresh_token to work.
BUT, if I attempt to refresh the token from the client side, I get invalid user credentials.
I just double checked. I am fully logged in with jwt and jwt_refresh. I still cannot refresh the token from the client side.
I'm not sure what's going on. My first thought is maybe a CORS thing?

**Reply by MrCasquette** [reply]
As for me it was just a concern about a JSON.parse. I just added it and it works. Hope it helps.

**Reply by irishburlybear** [reply]
And as for myself, It was the forgettable {browser} that saved me. And a function executing twice..

**Reply by orlsanchez** [reply]
I'm having something similar. I'm getting the 401 Invalid User Credentials response when running the Windows Desktop App. Authentication works well in Chrome and Edge. It was working with the Desktop app at the beginning, but stopped after a while. It looks like it has something to do with the refresh token, but haven't been able to figure it out. Any help is appreciated.

**Comment by irishburlybear** [comment]
Self Solved...

**Reply by royemosby** [reply]
What ended up being your resolution? I'm having something similar happen to me in Angular service.

**Comment by Pachat** [comment]
Still getting this error with 9.22.4 either under Edge and Firefox, either locally and remotely when disconnected.
`/auth/refresh 401 (Unauthorized)`

**Comment by norricorp** [comment]
I am also getting this problem with SDK auth.refresh(). And during auth.logout(), though it works, browser console shows a refresh error. This is with Firefox. Refresh token is in cookie.
For those who have solved it, can you give more details?

**Comment by mjrussell91** [comment]
I turned debug logging on for Directus and the error is being thrown at `./node_modules/@directus/api/dist/services/authentication.js` line 233 indicating this if statement:

`if (!record || (!record.user_id && !record.share_id)) {
    throw new InvalidCredentialsException();
}`

I console logged `record` and found that while I had a `user_id` I did not have a `share_id`. I'm not familiar with the Directus shares data, but it seems like either the knex join is failing, the user data is missing a share_id, or this condition is incorrect (although removing it didn't return successfully either).

`record`:
`{
  session_expires: 2023-05-30T10:52:16.572Z,
  user_id: '<redacted>',
  user_first_name: 'Admin',
  user_last_name: 'User',
  user_email: '<redacted>',
  user_password: '<redacted>',
  user_status: 'active',
  user_provider: 'default',
  user_external_identifier: null,
  user_auth_data: null,
  role_id: '<redacted>',
  role_admin_access: true,
  role_app_access: true,
  share_id: null,
  share_item: null,
  share_role: null,
  share_collection: null,
  share_start: null,
  share_end: null,
  share_times_used: null,
  share_max_uses: null
}`

I don't have a solution but maybe this is a bug and not poor configuration. I feel like I'm correctly posting the refresh token to the endpoint according to the docs, but receiving 401 Unauthorized in response.

**Reply by mjrussell91** [reply]
Actually I managed to resolve the 401 Unauthorized error by changing the default HOST in the Directus .env to "localhost" instead of "0.0.0.0" as the cookie's domain was set to localhost when hitting the refresh endpoint which is a conflict.

**Reply by drennvinn** [reply]
thanks! it worked for me too
